### ISSUE TITLE
Hypertext floating images are always at the top of paragraphs

### ISSUE THREAD

**Issue by v-rob** [minetest/minetest#9328]
##### Minetest version
```
Minetest 5.2.0-dev-e5a96d42
```

##### OS / Hardware
Operating system: Linux Mint 19.1

##### Summary
In the formspec `hypertext` tag, using the `float` attribute on paragraphs makes the images always appear at the top of the paragraph wherever the image is places.  With more than one image, they are stacked on top of each other.

Where the `<img>` tags are is shown in the text by the `[ignore][/ignore]` and `[air][/air]` for the ignore and air images, respectively.
![image](https://user-images.githubusercontent.com/31123645/72703231-ec189580-3b09-11ea-8b14-89a6329d483a.png)

##### Steps to reproduce
```lua
"formspec_version[3]size[20,12]hypertext[0.5,0.5;19,11;name;".. minetest.formspec_escape(
[=[<global margin=15% background=#FFFFFF55><style size=1> </style>
<bigger>Test Page</bigger>
<big>Subheading</big><justify><tag name=action color=skyblue hovercolor=lightblue>Text <b><action name=x>Action</action></b> Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text [ignore]<img name=ignore.png width=48 float=right>[/ignore]Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text [air]<img name=air.png width=48 float=right>[/air]Text Text Text Text Text Text Text Text Text Text Text Text</justify>]=]) .."]"..
```

**Comment by pyrollo** [comment]
Hi v-rob,

I'm not surprised with this behavior. The Hypertext element was designed to improve textareas and labels but it is still very rudimentary.

Floating images are anchored to paragraphs and it would be preferable (for now) to have only one floating image per paragraph (It is a limitation but it is yet much better than former text areas to display things).

There could be many different ways to manage multiple floating images but they implies to make choices.
  * Should image tag split paragraph in two ?
  * Should floating images be displayed aside each other ?
  * Should floating images be displayed below others ?

In short: how should floating images behave ?

Hypertext goal was not to implement complete html specifications but just improve text displaying in formspecs. In my opinion, it should stay quite simple (it is yet a bit complex).

**Comment by v-rob** [comment]
It does make sense that hypertext can be rudimentary, but I think we should still try to make it versatile.  I don't want hypertext to ever make the kinds of webpages we see nowadays; that's ridiculous, and if we want that, we'll embed Chrome in Minetest.  But I do think that it should have the ability to display text and images the way they are meant to by the author.

Well, here's how I think that floating images should work (black is paragraph, red line is `<img>` tag, and red rectangle is the floating image):

![floating](https://user-images.githubusercontent.com/31123645/72758506-dcd82d00-3b87-11ea-88db-ab13d877eaa4.png)

1. The simplest floating image.  The top of the floating image is aligned with the img tag
2. If they are far enough apart, display multiple in the same paragraph like 1.
3. When too close to be separate, push the next ones down below the first.
4. Now my plans diverge.  Once we have too many images to fit below the first image, what do we do?
    1. Start to extend into the next paragraph.
    2. Make space for the rest of the images before making the next paragraph.

When we choose either `i` or `ii`, consider this: `ii` will look kind of bad with tiny one to two line paragraphs with a floating image, but it might be easier to implement.  `i` will be more consistent with how a floating image is supposed to work (detached from everything else, not exclusively part of anything).  I would choose `i` if I could.

When making this, I kept this idea in mind: hypertext parses from the top down, so the first images forcing everything else down makes this much easier to implement than say, averaging the two conflicting images.

To implement multiple floating images, I suggest that, while parsing, place the floating image immediately before continuing with text parsing.  Then, you can find how wide the margins are including the floating images with some helper function while rendering a line of text.

**Comment by pyrollo** [comment]
I'm not sure to agree with "1 - The top of the floating image is aligned with the img tag"
Maybe it's good but it is yet a bit complex in some cases. 

Imagine the IMG tag is at the end of the line, when future image placing overlaps : 

![image](https://user-images.githubusercontent.com/13189280/72813697-167e6600-3c64-11ea-8af3-4f4b791655da.png)

A: I find a IMG tag so image should be aligned to the top of it.
B: Once image inserted, IMG tag is pushed next line. But image should be aligned to this tag.
C: Once image at its new position, IMG tag is back on first line. :( still not properly aligned.

The simplest choice is to keep B as final result even if its not good (but if I remember correctly, it will imply to start drawing line, then go back to restart with a different width).

I feel (not yet dug enough) that many similar cases will rise up if we want floating image to be aligned with image tag (I did not test with different font sizes and inline images, it would be a nightmare).

I also feel that the best to do is to have images anchored to paragraphs (even if it forces to split some of them for some placings) and have a good mechanism to avoid image stacking.

Sorry, these are only feelings for now, but I wanted to share them before going further.

**Comment by v-rob** [comment]
This is a good point.  Using a web browser (and assuming we have `clear: both` set for images, it would be worse to implement otherwise), B is what happens, but I think C is probably easier and better for us for ease of implementation.  Now, as for working with different font sizes and inline images, we just check the margins on that line.

But if this really is too infeasible, I won't press it.  I do think that this should be fixed sometime, but I don't expect it to be anytime soon.

**Comment by pyrollo** [comment]
I totally agree for tables being more important.
I guess the good plan for this issue, is just avoid images stacking, using the most simple way, which ever it is.
